File: focus_todo_mobile/lib/sync/fake_sync_transport.dart
import '../shared/models/task.dart';
import '../shared/models/task_sync_metadata.dart';
import 'canonical_task_mapper.dart';
import 'sync_transport.dart';
import 'merge_engine.dart';
import 'mock_sync_flow.dart';

class FakeSyncTransport implements SyncTransport {
  FakeSyncTransport({
    List<CanonicalTask> initialTasks = const <CanonicalTask>[],
  }) {
    for (final CanonicalTask task in initialTasks) {
      _remoteTasks[task.id] = _clone(task);
    }
  }

  final Map<String, CanonicalTask> _remoteTasks = <String, CanonicalTask>{};
  String? _nextPushError;
  String? _nextPullError;

  void failNextPush(String message) {
    _nextPushError = message;
  }

  void failNextPull(String message) {
    _nextPullError = message;
  }

  List<CanonicalTask> snapshot() {
    final tasks = _remoteTasks.values.map(_clone).toList(growable: false)
      ..sort(
        (CanonicalTask left, CanonicalTask right) =>
            left.id.compareTo(right.id),
      );
    return tasks;
  }

  @override
  Future<void> pushTasks(List<CanonicalTask> outboundBatch) async {
    if (_nextPushError != null) {
      final message = _nextPushError!;
      _nextPushError = null;
      throw Exception(message);
    }

    for (final CanonicalTask outboundTask in outboundBatch) {
      final existing = _remoteTasks[outboundTask.id];
      if (existing == null) {
        _remoteTasks[outboundTask.id] = _clone(outboundTask);
        continue;
      }

      final merged = mergeTask(
        fromCanonicalTask(existing),
        fromCanonicalTask(outboundTask),
      );
      _remoteTasks[outboundTask.id] = toCanonicalTask(merged);
    }
  }

  @override
  Future<List<CanonicalTask>> pullTasks() async {
    if (_nextPullError != null) {
      final message = _nextPullError!;
      _nextPullError = null;
      throw Exception(message);
    }

    return snapshot();
  }

  @override
  Future<List<CanonicalTask>> syncOnce(
    List<Task> currentTasks,
    List<TaskSyncMetadata> currentSyncMetadata,
  ) async {
    final coordinatorState = initializeMockSyncState(
      tasks: currentTasks,
      syncMetadata: currentSyncMetadata,
    );
    final outboundBatch = prepareMockOutboundPreview(coordinatorState);

    await pushTasks(outboundBatch);
    return pullTasks();
  }

  CanonicalTask _clone(CanonicalTask task) {
    return CanonicalTask(
      schemaVersion: task.schemaVersion,
      id: task.id,
      userId: task.userId,
      title: task.title,
      completed: task.completed,
      createdAt: task.createdAt,
      updatedAt: task.updatedAt,
      createdByDeviceId: task.createdByDeviceId,
      updatedByDeviceId: task.updatedByDeviceId,
      note: task.note,
      dueDate: task.dueDate,
      priority: task.priority,
      category: task.category,
      completedAt: task.completedAt,
      deletedAt: task.deletedAt,
    );
  }

  @override
  Future<SyncWriteResult> writeTask(
    CanonicalTask task,
    String? expectedUpdatedAt,
  ) async {
    throw UnimplementedError(
      'FakeSyncTransport does not support versioned writeTask; use syncOnce in fake flows',
    );
  }
}

File: focus_todo_computer/focus-todo/src/sync/fakeSyncTransport.ts
import type { Task, TaskSyncMetadata } from '../types/task'
import { toCanonicalTask, fromCanonicalTask, type CanonicalTask } from './canonicalTaskMapper'
import { mergeTask } from './mergeEngine'
import { syncCoordinator } from './syncCoordinator'
import type { SyncTransport } from './syncTransport'

export class FakeSyncTransport implements SyncTransport {
  private readonly remoteTasks = new Map<string, CanonicalTask>()
  private nextPushError: string | null = null
  private nextPullError: string | null = null

  constructor(initialTasks: CanonicalTask[] = []) {
    for (const task of initialTasks) {
      this.remoteTasks.set(task.id, { ...task })
    }
  }

  failNextPush(message: string): void {
    this.nextPushError = message
  }

  failNextPull(message: string): void {
    this.nextPullError = message
  }

  snapshot(): CanonicalTask[] {
    return [...this.remoteTasks.values()]
      .map((task) => ({ ...task }))
      .sort((left, right) => left.id.localeCompare(right.id))
  }

  async pushTasks(outboundBatch: CanonicalTask[]): Promise<void> {
    if (this.nextPushError) {
      const message = this.nextPushError
      this.nextPushError = null
      throw new Error(message)
    }

    for (const outboundTask of outboundBatch) {
      const existing = this.remoteTasks.get(outboundTask.id)
      if (!existing) {
        this.remoteTasks.set(outboundTask.id, { ...outboundTask })
        continue
      }

      const merged = mergeTask(
        fromCanonicalTask(existing),
        fromCanonicalTask(outboundTask),
      )
      this.remoteTasks.set(outboundTask.id, toCanonicalTask(merged))
    }
  }

  async pullTasks(): Promise<CanonicalTask[]> {
    if (this.nextPullError) {
      const message = this.nextPullError
      this.nextPullError = null
      throw new Error(message)
    }

    return this.snapshot()
  }

  async syncOnce(
    currentTasks: Task[],
    currentSyncMetadata: TaskSyncMetadata[],
  ): Promise<CanonicalTask[]> {
    const outboundBatch = syncCoordinator.prepareOutboundBatch(
      syncCoordinator.initializeAfterHydration({
        tasks: currentTasks,
        syncMetadata: currentSyncMetadata,
      }),
    )

    await this.pushTasks(outboundBatch)
    return this.pullTasks()
  }
}
